File: Plantemon/app/src/main/java/com/g4ng/logic/BattleHandler.java
package com.g4ng.logic;

import android.util.Log;

import com.g4ng.model.Plant;
import com.g4ng.model.Player;
import com.g4ng.model.BattleState;

import java.util.Objects;
import java.util.ArrayList;
import java.util.List;

public class BattleHandler {
    private static final String TAG = "BattleHandler";

    private Player player1;
    private Player player2;

    // The Strategies
    private BattleController p1Controller;
    private BattleController p2Controller;

    private Action p1SelectedAction;
    private Action p2SelectedAction;
    private BattleState state;
    protected List<String> turnResults = new ArrayList<>();
    public BattleHandler(Player player1, Player player2, BattleController p1Controller, BattleController p2Controller) {
        this.player1 = Objects.requireNonNull(player1);
        this.player2 = Objects.requireNonNull(player2);
        this.p1Controller = Objects.requireNonNull(p1Controller);
        this.p2Controller = Objects.requireNonNull(p2Controller);
        this.state = BattleState.P1_MOVE;

        Log.d(TAG, "Battle started: " + player1.getUsername() + " vs " + player2.getUsername());
        advanceState();
    }

    public void applyAction(Player player, Action action) {
        if (player == player1) {
            p1SelectedAction = action;
        } else if (player == player2) {
            p2SelectedAction = action;
        }
        advanceState();
    }

    public void advanceState() {
        if (state == BattleState.END) return;

        if (checkWin()) {
            state = BattleState.END;
            handleEnd();
            return;
        }

        switch (state) {
            case P1_MOVE:
                if (p1SelectedAction == null) {
                    // Ask P1's controller for a move
                    p1Controller.requestAction(this, player1);
                } else {
                    state = BattleState.P2_MOVE;
                    advanceState();
                }
                break;

            case P2_MOVE:
                if (p2SelectedAction == null) {
                    // Ask P2's controller for a move
                    p2Controller.requestAction(this, player2);
                } else {
                    state = BattleState.PROCESSING;
                    advanceState();
                }
                break;

            case PROCESSING:
                processTurn();
                resetRound();
                state = BattleState.P1_MOVE;
                advanceState(); // Loop back to P1
                break;
        }
    }
    public List<String> getLatestTurnResults() {
        return turnResults;
    }

    public void processTurn() {
        turnResults.clear();
        Plant p1 = player1.getCurrentPlant();
        Plant p2 = player2.getCurrentPlant();

        if (p1.getSpeed() >= p2.getSpeed()) {
            turnResults.add("--- " + p1.getName() + " is faster! ---");
            executeSequence(p1SelectedAction, p2SelectedAction, player1, player2);
        } else {
            turnResults.add("--- " + p2.getName() + " is faster! ---");
            executeSequence(p2SelectedAction, p1SelectedAction, player2, player1);
        }

        updatePlayers();
        advanceState();
    }

    public void updatePlayers(){
        System.out.println("Status Update:");
        System.out.println(player1.getUsername() + "'s " + player1.getCurrentPlant().getName() +
                ": " + player1.getCurrentPlant().getCurrentHealth() + " HP");
        System.out.println(player2.getUsername() + "'s " + player2.getCurrentPlant().getName() +
                ": " + player2.getCurrentPlant().getCurrentHealth() + " HP");
    }

    protected void executeSequence(Action firstAction, Action secondAction, Player firstPlayer, Player secondPlayer) {
        Action firstPlayerAction = (firstPlayer == player1) ? p1SelectedAction : p2SelectedAction;
        Action secondPlayerAction = (secondPlayer == player1) ? p1SelectedAction : p2SelectedAction;

        if (firstAction != null && !firstPlayer.getCurrentPlant().isDead()) {
            turnResults.add(firstAction.execute(firstPlayer, secondPlayer, secondPlayerAction));
        }

        if (secondPlayer.getCurrentPlant().isDead()) {
            turnResults.add(secondPlayer.getUsername() + "'s " + secondPlayer.getCurrentPlant().getName() + " fainted!");
            return;
        }

        if (secondAction != null && !firstPlayer.getCurrentPlant().isDead()) {
            turnResults.add(secondAction.execute(secondPlayer, firstPlayer, firstPlayerAction));

            if (firstPlayer.getCurrentPlant().isDead()) {
                turnResults.add(firstPlayer.getUsername() + "'s " + firstPlayer.getCurrentPlant().getName() + " fainted!");
            }
        }
    }

    public boolean checkWin() {
        return player1.getCurrentPlant().isDead() || player2.getCurrentPlant().isDead();
    }

    protected void handleEnd() {
        if (player1.getCurrentPlant().isDead()) {
            turnResults.add(player2.getUsername() + " wins!");
        } else {
            turnResults.add(player1.getUsername() + " wins!");
        }
        player1.restoreGarden();
        player2.restoreGarden();
    }

    public void resetRound() {
        p1SelectedAction = null;
        p2SelectedAction = null;
    }

    public BattleState getState() {
        return state;
    }
}

File: plantemon-web/src/lib/domain/battle-handler.ts
import { BattleState } from "./types";
import type { Action, BattleController } from "./types";
import type { Player } from "./player";
import { requireCurrentPlant } from "./require-plant";

/**
 * Port of logic/BattleHandler.java.
 *
 * Drives a two-player turn loop through a small state machine. advanceState()
 * is re-entrant by design: it recurses until it either lands on a state that
 * needs input (a HumanController simply returns, parking the battle) or reaches
 * END. Both players lock in an Action, then PROCESSING resolves them in speed
 * order.
 */
export class BattleHandler {
  private readonly player1: Player;
  private readonly player2: Player;
  private readonly p1Controller: BattleController;
  private readonly p2Controller: BattleController;

  private p1SelectedAction: Action | null = null;
  private p2SelectedAction: Action | null = null;
  private state: BattleState = BattleState.P1_MOVE;

  protected turnResults: string[] = [];

  constructor(
    player1: Player,
    player2: Player,
    p1Controller: BattleController,
    p2Controller: BattleController,
  ) {
    this.player1 = player1;
    this.player2 = player2;
    this.p1Controller = p1Controller;
    this.p2Controller = p2Controller;
    this.advanceState();
  }

  applyAction(player: Player, action: Action | null): void {
    if (player === this.player1) {
      this.p1SelectedAction = action;
    } else if (player === this.player2) {
      this.p2SelectedAction = action;
    }
    this.advanceState();
  }

  advanceState(): void {
    if (this.state === BattleState.END) return;

    if (this.checkWin()) {
      this.state = BattleState.END;
      this.handleEnd();
      return;
    }

    switch (this.state) {
      case BattleState.P1_MOVE:
        if (this.p1SelectedAction === null) {
          this.p1Controller.requestAction(this, this.player1);
        } else {
          this.state = BattleState.P2_MOVE;
          this.advanceState();
        }
        break;

      case BattleState.P2_MOVE:
        if (this.p2SelectedAction === null) {
          this.p2Controller.requestAction(this, this.player2);
        } else {
          this.state = BattleState.PROCESSING;
          this.advanceState();
        }
        break;

      case BattleState.PROCESSING:
        this.processTurn();
        this.resetRound();
        this.state = BattleState.P1_MOVE;
        this.advanceState();
        break;
    }
  }

  /** Log lines produced by the most recent turn, for the battle log UI. */
  getLatestTurnResults(): string[] {
    return this.turnResults;
  }

  processTurn(): void {
    this.turnResults.length = 0;
    const plant1 = requireCurrentPlant(this.player1);
    const plant2 = requireCurrentPlant(this.player2);

    // Ties go to player 1.
    if (plant1.getSpeed() >= plant2.getSpeed()) {
      this.turnResults.push(`--- ${this.player1.getUsername()}'s ${plant1.getName()} is faster! ---`);
      this.executeSequence(this.p1SelectedAction, this.p2SelectedAction, this.player1, this.player2);
    } else {
      this.turnResults.push(`--- ${this.player2.getUsername()}'s ${plant2.getName()} is faster! ---`);
      this.executeSequence(this.p2SelectedAction, this.p1SelectedAction, this.player2, this.player1);
    }
  }

  /**
   * Resolves both actions in order. Each action receives the opponent's action
   * so it can read its defense value.
   *
   * The Java version re-derived these two actions from p1SelectedAction /
   * p2SelectedAction via an identity check on the player; that always resolved
   * to firstAction and secondAction at both call sites, so they are used
   * directly here.
   */
  protected executeSequence(
    firstAction: Action | null,
    secondAction: Action | null,
    firstPlayer: Player,
    secondPlayer: Player,
  ): void {
    if (firstAction !== null && !requireCurrentPlant(firstPlayer).isDead()) {
      this.turnResults.push(firstAction.execute(firstPlayer, secondPlayer, secondAction));
    }

    if (requireCurrentPlant(secondPlayer).isDead()) {
      const fainted = requireCurrentPlant(secondPlayer);
      this.turnResults.push(`${secondPlayer.getUsername()}'s ${fainted.getName()} fainted!`);
      return;
    }

    if (secondAction !== null && !requireCurrentPlant(firstPlayer).isDead()) {
      this.turnResults.push(secondAction.execute(secondPlayer, firstPlayer, firstAction));

      if (requireCurrentPlant(firstPlayer).isDead()) {
        const fainted = requireCurrentPlant(firstPlayer);
        this.turnResults.push(`${firstPlayer.getUsername()}'s ${fainted.getName()} fainted!`);
      }
    }
  }

  checkWin(): boolean {
    return requireCurrentPlant(this.player1).isDead() || requireCurrentPlant(this.player2).isDead();
  }

  protected handleEnd(): void {
    if (requireCurrentPlant(this.player1).isDead()) {
      this.turnResults.push(`${this.player2.getUsername()} wins!`);
    } else {
      this.turnResults.push(`${this.player1.getUsername()} wins!`);
    }
    this.player1.restoreGarden();
    this.player2.restoreGarden();
  }

  resetRound(): void {
    this.p1SelectedAction = null;
    this.p2SelectedAction = null;
  }

  getState(): BattleState {
    return this.state;
  }
}
